### PR TITLE
Python bindings fix ups in preparation to Sparse Tensor introduction

### PR DESCRIPTION
**Pull request by yuslepukhin**
**Description**:
Fix up constness in pybindings
  Fix up return argument treatments.
  Specifically, for all functions that return pointers or references
  to the members of other pybind registered classes, we want not to copy
  them, but internally bump up a reference to the hosting class so they do not
  disappear before the reference to the returned members is re-claimed.
  This policy is applied by default to def_property and def_readwrite but not to def_readonly
  and other def methods.
  See https://pybind11-jagerman.readthedocs.io/en/stable/advanced.html#return-value-policies
  https://pybind11.readthedocs.io/en/stable/advanced/functions.html#return-value-policies

**Motivation and Context**
We are hitting obj limit size, we also want to improve code organization.

### CODE REVIEW

File: onnxruntime/core/session/IOBinding.cc [hunk 1]
```diff
@@ -109,6 +109,8 @@ void IOBinding::ClearOutputs() {
 
 const std::vector<std::string>& IOBinding::GetOutputNames() const { return output_names_; }
 
+const std::vector<OrtValue>& IOBinding::GetOutputs() const { return outputs_; }
+
 std::vector<OrtValue>& IOBinding::GetOutputs() { return outputs_; }
```
Comment (hariharans29) [inline onnxruntime/core/session/IOBinding.cc:114]: do we need the method returning non-const outputs in the internal IOBinding class ? Under what circumstances do we need these two variants ? A comment may help someone trying to understand the code..

File: onnxruntime/python/onnxruntime_pybind_iobinding.cc [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+#include "core/framework/tensorprotoutils.h"
+#include "core/framework/TensorSeq.h"
+#include "core/session/IOBinding.h"
+
+namespace onnxruntime {
+namespace python {
+
+namespace py = pybind11;
+
+void addIOBindingMethods(pybind11::module& m) {
```
Comment (hariharans29) [inline onnxruntime/python/onnxruntime_pybind_iobinding.cc:24]: nit: function name style ?

File: onnxruntime/core/session/IOBinding.cc (hunk 1 above)
Comment (yuslepukhin) [reply onnxruntime/core/session/IOBinding.cc:114]: It is an overload.

File: onnxruntime/python/onnxruntime_pybind_ortvalue.cc [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        }
+
+        ORT_ENFORCE(type_proto != nullptr, "Unknown type of OrtValue: ", ort_value->Type());
+
+        return *ONNX_NAMESPACE::Utils::DataTypeUtils::ToType(*type_proto);
+      })
+      .def("is_tensor", [](const OrtValue* ort_value) -> bool {
+        return ort_value->IsTensor();
+      })
+      .def("is_sparse_tensor", [](const OrtValue* ort_value) -> bool {
```
Comment (hariharans29) [inline onnxruntime/python/onnxruntime_pybind_ortvalue.cc:183]: sorry if I am missing something very obvious - but how does a user create an OrtValue in Python that is a sparse tensor or a sequence tensor ? Looking at it, it looks like they can only get hold of such OrtValues, fromwhat is returned from IoBinding ?

Comment (yuslepukhin) [reply onnxruntime/python/onnxruntime_pybind_ortvalue.cc:183]: They will be returned from run call, although this is mostly coming in the next iteration, when OrtValue will contain SparseTensor. TensorSeq is actually supported now but apparently not via python

File: onnxruntime/test/python/onnxruntime_test_python_iobinding.py [hunk 4]
```diff
@@ -24,6 +24,14 @@ def create_expected_output(self):
     def create_expected_output_alternate(self):
         return np.array([[2.0, 8.0], [18.0, 32.0], [50.0, 72.0]], dtype=np.float32)
 
+    def test_ort_value_properties(self):
```
Comment (hariharans29) [inline onnxruntime/test/python/onnxruntime_test_python_iobinding.py:27]: If I am not mistaken, such a "generic" (non-IOBinding related) OrtValue test already exists in the regular python test file. This file as the name indicates is only for IOBinding related tests.

Comment (yuslepukhin) [reply onnxruntime/test/python/onnxruntime_test_python_iobinding.py:27]: I could not find it

File: onnxruntime/python/onnxruntime_pybind_state_common.h [hunk 5]
```diff
@@ -139,6 +328,20 @@ void InitializeSession(InferenceSession* sess,
 // Checks if PyErrOccured, fetches status and throws.
 void ThrowIfPyErrOccured();
 
+void addOrtValueMethods(pybind11::module& m);
```
Comment (hariharans29) [inline onnxruntime/python/onnxruntime_pybind_state_common.h:331]: nit: function naming style

File: onnxruntime/python/onnxruntime_pybind_state.cc [hunk 6]
```diff
@@ -2172,6 +1605,8 @@ PYBIND11_MODULE(onnxruntime_pybind11_state, m) {
 
   addGlobalMethods(m, env);
   addObjectMethods(m, env);
+  addOrtValueMethods(m);
```
Comment (hariharans29) [inline onnxruntime/python/onnxruntime_pybind_state.cc:1608]: nit: naming style

File: onnxruntime/python/onnxruntime_pybind_exceptions.h [hunk 7]
```diff
@@ -43,48 +45,9 @@ struct EPFail : std::runtime_error {
...
-  pybind11::register_exception<NoModel>(m, "NoModel");
-  pybind11::register_exception<EngineError>(m, "EngineError");
-  pybind11::register_exception<RuntimeException>(m, "RuntimeException");
-  pybind11::register_exception<InvalidProtobuf>(m, "InvalidProtobuf");
-  pybind11::register_exception<ModelLoaded>(m, "ModelLoaded");
-  pybind11::register_exception<NotImplemented>(m, "NotImplemented");
-  pybind11::register_exception<InvalidGraph>(m, "InvalidGraph");
-  pybind11::register_exception<EPFail>(m, "EPFail");
-}
+void RegisterExceptions(pybind11::module& m);
```
Comment (hariharans29) [inline onnxruntime/python/onnxruntime_pybind_exceptions.h:48]: is there value in we just moving these declarations into onnxruntime_pybind_state_common.h and removing this file altogether? The definitions live in onnxruntime_pybind_state_common.cc anyway ?

File: onnxruntime/test/python/onnxruntime_test_python_iobinding.py (hunk 4 above)
Comment (hariharans29) [reply onnxruntime/test/python/onnxruntime_test_python_iobinding.py:27]: https://github.com/microsoft/onnxruntime/blob/ff655175ff2ecb32f24cdc43870a6c34667d73d6/onnxruntime/test/python/onnxruntime_test_python.py#L836

File: onnxruntime/python/onnxruntime_pybind_exceptions.h (hunk 7 above)
Comment (yuslepukhin) [reply onnxruntime/python/onnxruntime_pybind_exceptions.h:48]: Discussed offline

**Review by hariharans29 — APPROVED** [review]

**Pull request merged**